onnxruntime/core/common/threadpool.cc

```diff
@@ -402,29 +402,45 @@ void ThreadPool::ParallelForFixedBlockSizeScheduling(const std::ptrdiff_t total,
     return;
   }
 
-  // Split the work across threads in the pool.  Each work item will run a loop claiming iterations,
-  // hence we need at most one for each thread, even if the numberof blocks of iterations is larger.
   auto d_of_p = DegreeOfParallelism(this);
-  auto num_blocks = total / block_size;
-  auto num_threads_inc_main = NumThreads() + 1;
-  int num_work_items = static_cast<int>(std::min(static_cast<std::ptrdiff_t>(num_threads_inc_main), num_blocks));
-  assert(num_work_items > 0);
-
-  LoopCounter lc(total, d_of_p, block_size);
-  std::function<void(unsigned)> run_work = [&](unsigned idx) {
-    unsigned my_home_shard = lc.GetHomeShard(idx);
-    unsigned my_shard = my_home_shard;
-    uint64_t my_iter_start, my_iter_end;
-    while (lc.ClaimIterations(my_home_shard, my_shard, my_iter_start, my_iter_end)) {
-      fn(static_cast<std::ptrdiff_t>(my_iter_start),
-         static_cast<std::ptrdiff_t>(my_iter_end));
-    }
-  };
-
-  // Run the work in the thread pool (and in the current thread).  Synchronization with helping
-  // threads is handled within RunInParallel, hence we can deallocate lc and other state captured by
-  // run_work.
-  RunInParallel(run_work, num_work_items, block_size);
+  if (thread_options_.dynamic_block_base_ <= 0) {
+    // Split the work across threads in the pool.  Each work item will run a loop claiming iterations,
+    // hence we need at most one for each thread, even if the number of blocks of iterations is larger.
+    auto num_blocks = total / block_size;
+    auto num_threads_inc_main = NumThreads() + 1;
+    int num_work_items = static_cast<int>(std::min(static_cast<std::ptrdiff_t>(num_threads_inc_main), num_blocks));
+    assert(num_work_items > 0);
+
+    LoopCounter lc(total, d_of_p, block_size);
+    std::function<void(unsigned)> run_work = [&](unsigned idx) {
+      unsigned my_home_shard = lc.GetHomeShard(idx);
+      unsigned my_shard = my_home_shard;
+      uint64_t my_iter_start, my_iter_end;
+      while (lc.ClaimIterations(my_home_shard, my_shard, my_iter_start, my_iter_end)) {
+        fn(static_cast<std::ptrdiff_t>(my_iter_start),
+           static_cast<std::ptrdiff_t>(my_iter_end));
+      }
+    };
+    // Run the work in the thread pool (and in the current thread).  Synchronization with helping
+    // threads is handled within RunInParallel, hence we can deallocate lc and other state captured by
+    // run_work.
+    RunInParallel(run_work, num_work_items, block_size);
+  } else {
```

Comment:
Main comment on the code: can we structure this so that the loop itself is the same, but we use a different form of LoopCounter?  We would have

StaticLoopCounter == What we have today
DynamicLoopCounter == This PR

We could then use a C++ template for the loop itself, parameterized by "T", and instantiated once with T=StaticLoopCounter, and once with T=DynamicLoopCounter.

Reply:
Refactor a bit on current LoopCounter which accepts a block_size on ClaimIterations.